src/test/util_string_tests.cpp

```diff
@@ -14,15 +14,34 @@ BOOST_AUTO_TEST_SUITE(util_string_tests)
 // Helper to allow compile-time sanity checks while providing the number of
 // args directly. Normally PassFmt<sizeof...(Args)> would be used.
 template <unsigned NumArgs>
-inline void PassFmt(util::ConstevalFormatString<NumArgs> fmt)
+inline void PassFmt(util::ConstevalFormatString<util::NoEndCheck, NumArgs> fmt)
 {
     // This was already executed at compile-time, but is executed again at run-time to avoid -Wunused.
     decltype(fmt)::Detail_CheckNumFormatSpecifiers(fmt.fmt);
 }
-template <unsigned WrongNumArgs>
+template <unsigned NumArgs>
+inline void PassFmtNewline(util::ConstevalFormatString<util::TrailingNewlineCheck, NumArgs> fmt)
+{
+    // This was already executed at compile-time, but is executed again at run-time to avoid -Wunused.
+    decltype(fmt)::Detail_CheckNumFormatSpecifiers(fmt.fmt);
+}
+template <unsigned WrongNumArgs, util::EndCheck end_check = util::NoEndCheck>
 inline void FailFmtWithError(std::string_view wrong_fmt, std::string_view error)
 {
-    BOOST_CHECK_EXCEPTION(util::ConstevalFormatString<WrongNumArgs>::Detail_CheckNumFormatSpecifiers(wrong_fmt), const char*, HasReason(error));
+    using WrongFmt = util::ConstevalFormatString<end_check, WrongNumArgs>;
+    BOOST_CHECK_EXCEPTION(WrongFmt::Detail_CheckNumFormatSpecifiers(wrong_fmt), const char*, HasReason(error));
+}
+
+BOOST_AUTO_TEST_CASE(ConstevalFormatString_Newline)
+{
+    PassFmtNewline<0>("\n");
+    PassFmtNewline<1>("%s\n");
+    PassFmtNewline<0>("%%\n");
+    PassFmtNewline<1>("%\n"); // tinyformat will use the newline as format specifier
+
+    auto err_newline{"Format string must end with a newline"};
+    FailFmtWithError<1, util::TrailingNewlineCheck>("", err_newline);
```

Comment:
given that `PassFmtNewline<1>("\n")` would already fail - and this is an test for a missing newline, not invalid parameter count -, consider the following:
```suggestion
    FailFmtWithError<0, util::TrailingNewlineCheck>("", err_newline);
```